File: day4/day4.py


def is_non_decrease_and_adjecent(number):
    adjecent_flag = False
    for first, second in zip(number[:-1], number[1:]):
        if first > second:
            return False and adjecent_flag
        elif first == second:
            adjecent_flag = True

    return True and adjecent_flag

def is_non_decrease_and_adjecent_non_matching(number):
    adjecent_group = {}
    for first, second in zip(number[:-1], number[1:]):
        if first > second:
            return False 
        elif first == second:
            if second in adjecent_group.keys():
                adjecent_group[second] += 1
            else:
                adjecent_group[second] = 2

    return True and (2 in adjecent_group.values())


def password_combinations(password_range):
    ans = 0
    for password in range(password_range[0], password_range[1]):
        if is_non_decrease_and_adjecent(str(password)):
            ans += 1
    return ans

def password_combinations_non_matching(password_range):
    ans = 0
    for password in range(password_range[0], password_range[1]):
        if is_non_decrease_and_adjecent_non_matching(str(password)):
            ans += 1
    return ans


print(password_combinations([172930,683082]))
print(password_combinations_non_matching([172930,683082]))

File: day4.cpp
#include <iostream>
#include <map>

using namespace std;


bool is_non_decrease_and_adjecent(const string& number){
    bool adjecent_flag = false;
    for (auto first = number.begin(); first != number.end()-1; first++){
        string::const_iterator second = first + 1;
        if (*first > *second){
            return false;
        }
        else if (*first == *second){
            adjecent_flag = true;
        }

    }
    return true && adjecent_flag;
 }



bool is_non_decrease_and_adjecent_non_matching(const string& number){
    map<char, int> adjecent_group;
    bool adjecent_flag = false;
    for (auto first = number.begin(); first != number.end()-1; first++){
        string::const_iterator second = first + 1;
        if (*first > *second){
            return false;
        }
        else if (*first == *second){
            if (adjecent_group.count(*second)){
                adjecent_group[*second] += 1;
            }
            else {
                 adjecent_group[*second] = 2;   
            }
                
        }
    }
    for (const auto i: adjecent_group){
        if (i.second == 2){
            adjecent_flag =true;
        }
    }
    return true && adjecent_flag;
}

int main(){
    int ans1 = 0;
    int ans2 = 0;
    for (int i=172930; i<683083; i++){
        if (is_non_decrease_and_adjecent(to_string(i))){
            ans1 += 1;
        }
        if (is_non_decrease_and_adjecent_non_matching(to_string(i))){
            ans2 += 1;
        }
    }

    cout << ans1 << endl;
    cout << ans2 << endl;

    return 0;
}
